### PR TITLE
Fixes for new simulation check

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
Follow up to #4538. This PR finally makes use of the features introduces in the previous 2 PRs to address the limitations of the simulation approach. It also addresses 2 other smaller issues with the current code.

# Changes
1. We now use the state override errors and the simulation report to infer whether our state override errors caused the simulation to fail. If everything worked fine until the simulation hits the buy token transfer AND we previously failed to compute the necessary state overrides we now err on the side of allowing the order to be placed. The most relevant examples of this would be orders buying `stETH` or `xStocks` tokens as those are relatively popular and are currently not supported by our balance override code.
2. Most of the simulation disagreements I saw were caused by the new simulation filling the order more than possible. The currently active logic tries to send only 1, 10, and 100 wei of the sell token into the settlement contract (for partially fillable orders). To allow the same partially funded orders to be placed via the new logic we now chose the fill amount as `clamp(1 wei, sellToken.balanceOf(trader), full order amount)`.
3. Lastly we now also pipe the `full_balance_check` into the simulation setup which the user can opt-in to when they only want fully funded orders to be placed.

Additionally this PR also starts logging the order simulation report to help us debug the remaining cases of disagreements faster.

## How to test
Added a unit test for detecting and handling those edge cases of our simulation.

### CODE REVIEW

File: crates/shared/src/order_creation_simulation.rs [hunk 1]
```diff
@@ -65,13 +75,49 @@ impl OrderSimulating for OrderCreationSimulator {
         &self,
         order: &Order,
         full_app_data: &str,
+        full_balance_check: bool,
     ) -> Result<(), OrderSimulationError> {
+        let inputs = self
+            .prepare_simulation(order, full_app_data, full_balance_check)
+            .await?;
+        analyze_simulation(order, inputs).await
```
Comment (squadgazzz) [inline crates/shared/src/order_creation_simulation.rs:91]: This now uses `debug_trace_call` for every order instead of `eth_call`, right? Should we keep `simulate()` as the fast path and call `simulation_report()` only when it reverts?

Comment (MartinquaXD) [reply crates/shared/src/order_creation_simulation.rs:91]: I was on the fence about this one. Probably best to assume that the majority of orders are well behaved so falling back to the new logic only for reverts is not that costly.

File: crates/shared/src/order_creation_simulation.rs [hunk 2]
```diff
@@ -65,13 +75,49 @@ impl OrderSimulating for OrderCreationSimulator {
...
+                    .unwrap_or_default();
+                let clamped_sell_amount =
+                    sell_token_balance.clamp(U256::ONE, order.data.sell_amount);
+                let final_amount = match order.data.kind {
+                    OrderKind::Sell => clamped_sell_amount,
+                    // convert maxium sell amount to buy tokens
+                    OrderKind::Buy => clamped_sell_amount
+                        .checked_mul(order.data.buy_amount)
+                        .and_then(|val| val.checked_div(order.data.sell_amount))
+                        .unwrap_or(U256::ONE),
```
Comment (squadgazzz) [inline crates/shared/src/order_creation_simulation.rs]: `unwrap_or(U256::ONE)` only fires on `None`, which here means overflow in `checked_mul` or `sell_amount == 0` in `checked_div`. A small numerator is still a valid `Some(0)`, since integer division floors, so it slips past `unwrap_or` and we get `ExecutionAmount::Explicit(0)`. 

e.g.: clamped sell 1, buy 100, sell 1000 gives 100 / 1000 = `Some(0)`. `executed_amount` passes that through, so an underfunded partially-fillable buy order fills 0, the transfers are 0 and pass, and it clears the check without a real transfer. A `.max(U256::ONE)` on the final amount would keep the minimal fill. Does it make sense?

Comment (AryanGodara) [reply crates/shared/src/order_creation_simulation.rs]: +1 on this. A trailing .max(U256::ONE) at the end catches both this and any future overflow case in one place.
Worth a unit test in extract_critical_revert too, prepare_simulation isn't covered there rn.

Comment (MartinquaXD) [reply crates/shared/src/order_creation_simulation.rs]: Seems like an exotic edge case I'm not sure we'll ever hit but your comment is definitely correct and the fix is trivial. Applying it. 👍

**Review by AryanGodara — APPROVED** [review]
